Release web 2023 01 01

rohinisharma opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [public repo review queue], [private repo review queue]  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue]  

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [ ] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [X] Other, please clarify:
    - Adding public usages API with quota type parameter for onboarding onto QMS. This API has been implemented by other RPs such as Microsoft.Compute already.

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [X ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ X] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[public repo review queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AWaitForARMFeedback+-label%3AIDCDevDiv+draft%3Afalse+sort%3Acreated-asc
[private repo review queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AWaitForARMFeedback+-label%3AIDCDevDiv+draft%3Afalse+sort%3Acreated-asc
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AIDCDevDiv+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AIDCDevDiv+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

Commit:
Adding locations/usages endpoint and example

Commit:
Revert "Adding locations/usages endpoint and example"

This reverts commit 4a1110765d08c5c96d65f884237313515d842586.

Commit:
Adding back usages changes with up to date branch

Commit:
Updates based on verification tools

Commit:
Fixing spacing, adding missing comma

Commit:
ran prettier-fix for formatting

Commit:
Update for new QMS requirements with ZR

Commit:
Using prettier to fix linter error

Commit:
Adding ZR endpoint

Commit:
Fixing duplicate operationId

Commit:
Merge branch 'release-web-2023-01-01' of github.com:rohinisharma/azure-rest-api-specs into release-web-2023-01-01

Commit:
QMS Usages

zizw123 commented:
Please take a look at the lintDiff error, for list operation you should have pagination

Commit:
Validation changes

rohinisharma commented:
@zizw123 Can you take a look now, I have made changes based on your comment. Thanks!

Commit:
Updating schema

Commit:
Updating for collection

ms-zhenhua commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,54 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Web/locations/{location}/quotaType/{quotaType}/usages": {
```
Any resource type MUST implement GET for collection and single resource regardless if it's a singleton or not. 

Inconsistent or incomplete CRUD (https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review)

ms-zhenhua commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,54 @@
+      "get": {
+        "summary": "Gets subscription core usages for location, for a given type of quota.",
+        "description": "Gets the usage in cores for a subscription in a given location, for a specific quota type.",
+        "operationId": "GetUsagesInLocation",
```
should be end with _list:

https://github.com/Azure/azure-openapi-validator/blob/main/docs/list-in-operation-name.md

ms-zhenhua commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,54 @@
       }
     },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Web/locations/{location}/quotaType/{quotaType}/usages": {
+      "get": {
```
add x-ms-pageable for collection-get

https://github.com/Azure/azure-openapi-validator/blob/main/docs/pageable-operation.md

ms-zhenhua commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,54 @@
+        "operationId": "GetUsagesInLocation",
+        "parameters": [
+          {
+            "name": "location",
```
please use common type:

https://github.com/Azure/azure-rest-api-specs/blob/0ed6843ef8a11cc599c0a36233d4eaab5b77cccc/specification/common-types/resource-management/v5/types.json#L695C5-L703C7

Commit:
fixing pr comments

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
+          }
+        }
+      },
+      "x-ms-pageable": {
```
This x-ms-pageable is causing funny sounding linter failures because it is added at the wrong level. It should be part of the response definition. You've added it at the REST API verb level.

Commit:
Resolving comments

TimLovellSmith commented:
@rohinisharma 
With regard to our process, you've selected
'Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)' is only supposed to apply to preview api versions.

Isn't this api version already GA?

rohinisharma commented:
> @rohinisharma With regard to our process, you've selected 'Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)' is only supposed to apply to preview api versions.
> 
> Isn't this api version already GA? Why is it tagged as customer reported?

Updated process comment. I'm not sure why it's tagged as customer reported, I believe that label was auto added. I have removed it.

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
     },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Web/locations/{location}/quotaType/{quotaType}/usages": {
+      "get": {
+        "summary": "Gets subscription core usages for location, for a given type of quota.",
```
Consider rephrasing to "Gets the list of"?
At least, this ARM URI path looks intended to query for a list, according to ARM conventions.

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
+          },
+          {
+            "name": "quotaType",
+            "description": "The type of quota to get usages for",
```
What does a quota type string look like? How do people know they're using the right one?

rohinisharma replied:
For now quota type is "regional" or "zonal". We may add support for other types if our quota construct is expanded. We will return an appropriate error message for invalid quota type string.

TimLovellSmith replied:
I think you should define this as an enum so that its obvious to the API users what values to include.

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "./CommonDefinitions.json#/definitions/CsmUsageQuotaCollection"
```
You shouldn't be using the obsolete acronym CSM (the old internal name for ARM). Just call it 'UsageQuotaCollection'

TimLovellSmith replied:
[If this is a breaking change with GA APIs it may be simpler not to fix it though.]

rohinisharma replied:
There are other APIs that reference this definition, it will be simpler not to change

TimLovellSmith replied:
Okay, I guess we are stuck with it then. ;(

Commit:
Updating description

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
+      "get": {
+        "summary": "Gets the list of subscription core usages for all skus used in a location, for a given type of quota.",
+        "description": "Gets the list of usages in cores for all skus used by a subscription in a given location, for a specific quota type.",
+        "operationId": "GetUsagesInLocation_list",
```
I think they wouldn't want you to use both 'get' and 'list' in the operation name, but just choose one depending on whether its a 'get' or 'list' operation!

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -1492,6 +1539,15 @@
       "type": "string",
       "x-ms-parameter-location": "method"
+    },
+    "LocationParameter": {
```
Not sure why I don't see this as fixed when you resolved it. You still don't need to add locationParameter here when you can just reference common types one instead... do you?

TimLovellSmith commented:
Swagger APIView seems to be failing with
`Unhandled exception: System.Exception: Reference not found. $ref: #/definitions/userAssignedIdentity`
I wonder why...

Commit:
Resolving comments

rohinisharma commented:
> Swagger APIView seems to be failing with `Unhandled exception: System.Exception: Reference not found. $ref: #/definitions/userAssignedIdentity` I wonder why...

This definition is not referenced in any additions I made I don't think. Is there something I need to do to fix this?

TimLovellSmith commented:
The DeploymentLocations collection not having 'nextLink' is, in principle, a violation of ARM API guidelines.
Can we fix that? Or is there some special reason for omitting it?

rohinisharma commented:
> The DeploymentLocations collection not having 'nextLink' is, in principle, a violation of ARM API guidelines. Can we fix that? Or is there some special reason for omitting it?

I am not referencing DeploymentLocations collection in my changes. I understand this is a violation, but why is a fix needed in this PR? And was the issue not caught previously?

TimLovellSmith commented:
Can you please create a work item to fix the missing nextLink for DeploymentLocations in current or future api version?

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/examples/GetUsagesInLocation.json:
```diff
@@ -0,0 +1,27 @@
+    "200": {
+      "headers": {},
+      "body": {
+        "value": [
```
Why doesn't the payload of each 'usage' response contained in the list look like an ARM resource?

If you want a non-resource looking response payload, I am under the impression it has to be a POST method. [except for some special ARM APIS that don't follow the resource model conventions - this isn't one of those I presume?]

rohinisharma replied:
I am not returning an ARM resource but rather an object with information about the usage in a given region. Other RPs have implemented this method as a GET method already: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/labservices/resource-manager/Microsoft.LabServices/stable/2023-06-07/Usages.json

https://learn.microsoft.com/en-us/rest/api/compute/usage/list?tabs=HTTP

TimLovellSmith replied:
@rkmanda Right... sounds like this is becoming one of those instances of APIs where ARM guidelines weren't followed, and then it became a pattern... or am I just imagining some rules on GET responses that don't apply universally?

TimLovellSmith commented on specification/web/resource-manager/Microsoft.Web/stable/2023-01-01/ResourceProvider.json:
```diff
@@ -507,6 +507,53 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Web/locations/{location}/quotaType/{quotaType}/usages": {
```
With this resource type definition, I believe you also need to support 'point get', i.e. get a single usage by its name? At least, I hope to clarify this. Potential [ARMMustGuidelineViolation] /cc @rkmanda

rohinisharma replied:
I spoke with Zhenhua about this previously. This an API primarily for onboarding to QMS, and other RPs have previously implemented it, and did not have to create a point get method as well. Here is an example from Microsoft.LabServices: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/labservices/resource-manager/Microsoft.LabServices/stable/2023-06-07/Usages.json

TimLovellSmith replied:
@rohinisharma OK, I hope this is just a special case then. Its not an internal API right?

rohinisharma replied:
Correct, this is not an internal API.

ms-henglu commented:
Once you have addressed all feedback provided OR provided a rationale on why the feedback was not addressed, please remove the "ARMChangesRequested" label from the PR.

rohinisharma commented:
> Once you have addressed all feedback provided OR provided a rationale on why the feedback was not addressed, please remove the "ARMChangesRequested" label from the PR.

Hello! I have resolved the above comments as this API has been implemented by RPs previously as part of the onboarding process to Quota Management Service and I am following the same design.

kazrael2119 commented:
this js sdk breaking is not caused by this pr

qiaozha commented:
The JS SDK generation failed, we will need to investigate on this.

kazrael2119 commented:
your branch is not base on your latest dev-branch: ```release-web-2023-01-01```, please merge latest dev-branch.
and could you help to resolve this issue?https://github.com/Azure/azure-rest-api-specs/pull/24606#discussion_r1285479770
this pr was merged with js sdk generation failure. I guess this js sdk generation failure may the same issue after you merge latest dev-branch

rohinisharma committed:
Merge branch 'Azure:release-web-2023-01-01' into release-web-2023-01-01

rohinisharma commented:
> your branch is not base on your latest dev-branch: `release-web-2023-01-01`, please merge latest dev-branch. and could you help to resolve this issue?[#24606 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/24606#discussion_r1285479770) this pr was merged with js sdk generation failure. I guess this js sdk generation failure may the same issue after you merge latest dev-branch

I have merged the latest branch in. I am not sure what to do about the JS SDK failure, this was not introduced by my changes was it?

kazrael2119 commented:
> > your branch is not base on your latest dev-branch: `release-web-2023-01-01`, please merge latest dev-branch. and could you help to resolve this issue?[#24606 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/24606#discussion_r1285479770) this pr was merged with js sdk generation failure. I guess this js sdk generation failure may the same issue after you merge latest dev-branch
> 
> I have merged the latest branch in. I am not sure what to do about the JS SDK failure, this was not introduced by my changes was it?

can you help fix this issue? or can you ask someone related this change to fix this?

rohinisharma commented:
> > > your branch is not base on your latest dev-branch: `release-web-2023-01-01`, please merge latest dev-branch. and could you help to resolve this issue?[#24606 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/24606#discussion_r1285479770) this pr was merged with js sdk generation failure. I guess this js sdk generation failure may the same issue after you merge latest dev-branch
> > 
> > 
> > I have merged the latest branch in. I am not sure what to do about the JS SDK failure, this was not introduced by my changes was it?
> 
> can you help fix this issue? or can you ask someone related this change to fix this?

@kazrael2119 I have reached out to Sushmitha and she has made a PR to fix this: https://github.com/Azure/azure-rest-api-specs/pull/25493. Can we get it merged so that this PR can be unblocked?

kazrael2119 commented:
please merge your dev branch to update make js sdk generation work

rohinisharma committed:
Merge branch 'Azure:release-web-2023-01-01' into release-web-2023-01-01

rohinisharma commented:
> please merge your dev branch to update make js sdk generation work

@kazrael2119 JS SDK generation has succeeded

kazrael2119 commented:
this js sdk breaking should be non-breaking

rohinisharma commented:
/pr RequestMerge

sjanamma reviewed (approved):
approving for merge
